File: Go/Algorithmic_Blueprints/Sorting_and_Order_Statistics/heap_sort/heap_sort.go
package main

import "fmt"

// HeapSort sorts an array of integers using the heap sort algorithm.
func HeapSort(arr []int) {
	n := len(arr)

	// Build heap (rearrange array)
	for i := n/2 - 1; i >= 0; i-- {
		heapify(arr, n, i)
	}

	// One by one extract an element from heap
	for i := n - 1; i > 0; i-- {
		// Move current root to end
		arr[0], arr[i] = arr[i], arr[0]

		// call max heapify on the reduced heap
		heapify(arr, i, 0)
	}
}

func heapify(arr []int, n, i int) {
	largest := i
	l := 2*i + 1
	r := 2*i + 2

	if l < n && arr[l] > arr[largest] {
		largest = l
	}

	if r < n && arr[r] > arr[largest] {
		largest = r
	}

	if largest != i {
		arr[i], arr[largest] = arr[largest], arr[i]
		heapify(arr, n, largest)
	}
}

func main() {
	sample := []int{12, 11, 13, 5, 6, 7}
	fmt.Println("Original array:", sample)
	HeapSort(sample)
	fmt.Println("Sorted array:  ", sample)
}

File: python/algorithms/sorting/heap_sort.py

import heapq

def heap_sort_using_heapq(arr):
    """
    Sorts a list in ascending order using Python's built-in heapq module.

    Args:
        arr (list): A list of comparable elements.

    Returns:
        list: A new sorted list.
    """
    heap = arr.copy()
    heapq.heapify(heap)
    return [heapq.heappop(heap) for _ in range(len(heap))]


def heapify(arr, n, i):
    """
    Maintains the max-heap property for a subtree rooted at index i.

    Args:
        arr (list): The list representing the heap.
        n (int): Size of the heap.
        i (int): Index of the root element.
    """
    largest = i
    left = 2 * i + 1
    right = 2 * i + 2

    if left < n and arr[left] > arr[largest]:
        largest = left

    if right < n and arr[right] > arr[largest]:
        largest = right

    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)


def heap_sort(arr):
    """
    Sorts a list in ascending order using the Heap Sort algorithm.

    Args:
        arr (list): A list of comparable elements.

    Returns:
        list: The sorted list.
    """
    n = len(arr)

    # Build a max heap
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)

    # Extract elements one by one
    for i in range(n - 1, 0, -1):
        arr[0], arr[i] = arr[i], arr[0]
        heapify(arr, i, 0)

    return arr


def main():
    try:
        data = [2, 1, 4, 5, 3]

        sorted_with_heapq = heap_sort_using_heapq(data)
        sorted_with_manual = heap_sort(data.copy())

        print("Heapq Sort:", sorted_with_heapq)
        print("Manual Heap Sort:", sorted_with_manual)

    except Exception as e:
        print("Exception:", e)


if __name__ == "__main__":
    main()
